src/NovelRT/Graphics/GraphicsResourceManager.cpp

```diff
@@ -86,86 +139,138 @@ namespace NovelRT::Graphics
         stagingBuffer->UnmapAndWrite(vertexRegion);
         currentContext->Copy(vertexBuffer, stagingBuffer);
 
-        currentContext->EndFrame();
-        _graphicsDevice->Signal(currentContext->GetFence());
-        _graphicsDevice->WaitForIdle();
-        currentContext->BeginFrame();
-
         return vertexRegion;
     }
 
+    GraphicsMemoryRegion<GraphicsResource> GraphicsResourceManager::LoadIndexDataUntyped(void* data,
+                                                                                         size_t dataTypeSize,
+                                                                                         size_t dataLength,
+                                                                                         size_t alignment)
+    {
+        size_t sizeToStage = dataTypeSize * dataLength;
+        auto currentContext = _graphicsDevice->GetCurrentContext();
+
+        std::shared_ptr<GraphicsBuffer> stagingBuffer =
+            GetStagingBufferWithProperSizeHandling(sizeToStage, currentContext);
+
+        std::shared_ptr<GraphicsBuffer> indexBuffer =
+            GetOrCreateGraphicsBufferForAllocationSize(sizeToStage, GraphicsBufferKind::Index);
+        auto indexRegion = indexBuffer->Allocate(sizeToStage, alignment);
+        auto writeArea = stagingBuffer->Map<uint8_t>(indexRegion);
+#ifdef WIN32
+        memcpy_s(writeArea, sizeToStage, data, sizeToStage);
+#else
+        memcpy(writeArea, data, sizeToStage);
+#endif
+        stagingBuffer->UnmapAndWrite(indexRegion);
+        currentContext->Copy(indexBuffer, stagingBuffer);
+
+        return indexRegion;
+    }
+
     std::shared_ptr<GraphicsBuffer> GraphicsResourceManager::GetStagingBufferWithProperSizeHandling(
         size_t sizeToStage,
-        std::shared_ptr<GraphicsContext>& currentContext)
+        std::shared_ptr<GraphicsContext>& /*currentContext*/)
     {
-        std::shared_ptr<GraphicsBuffer> stagingBuffer = _stagingBuffer.getActual();
+        _stagingBufferSize = _stagingBufferSize < sizeToStage ? sizeToStage : _stagingBufferSize;
+
+        std::shared_ptr<GraphicsBuffer> chosenBuffer = nullptr;
+
+        for (auto&& buffer : _stagingBuffers[_contextIndex])
+        {
+            if (buffer->GetLargestFreeRegionSize() >= sizeToStage)
+            {
+                chosenBuffer = buffer;
```

Comment:
Can't we just return the buffer directly here, avoiding the null check later on?

Reply:
No because an adequate buffer might not be found.

Reply:
I'm not sure what you mean; if this code is executing then the buffer is not null, and it's also adequate enough, right?

Reply:
ah nevermind, I see what you mean - generally I don't like multiple return points but I can change it if you want, I just think it makes the code flow harder to understand.